File: src/main/java/j/chain_of_resposibility/inheritance/SpecialSupport.java
package j.chain_of_resposibility.inheritance;

import j.chain_of_resposibility.Trouble;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public final class SpecialSupport extends Support {
  private final int number; // この番号だけ解決できる

  public SpecialSupport(@NotNull String name, int number, @Nullable Support next) {
    super(name, next);
    this.number = number;
  }

  public SpecialSupport(@NotNull String name, int number) {
    this(name, number, null);
  }

  @Override
  protected boolean resolve(Trouble trouble) {
    return trouble.getNumber() == number;
  }
}

File: src/main/scala/s/chain_of_responsibility/inheritance/supports/SpecialSupport.scala
package s.chain_of_responsibility.inheritance.supports

import s.chain_of_responsibility.Trouble
import s.chain_of_responsibility.inheritance.Support

final class SpecialSupport(
    name: String,
    private val number: Int,
    next: Option[Support]
) extends Support(name, next) {
  def this(name: String, number: Int) = {
    this(name, number, None)
  }

  override protected def resolve(trouble: Trouble): Boolean =
    trouble.number == number
}
